components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodManager.kt

```diff
@@ -199,6 +205,105 @@ class GitpodManager : Disposable {
         }
     }
 
+    private val portsObserveJob = GlobalScope.launch {
+        if (application.isHeadlessEnvironment) {
+            return@launch
+        }
+
+        // Ignore ports that aren't actually used by the user (e.g. ports used internally by JetBrains IDEs)
+        val ignorePorts = listOf(5990, 5991, 6679, 6942, 6943)
+        val portsStatus = hashMapOf<Int, PortsStatus>()
+
+        val status = StatusServiceGrpc.newStub(supervisorChannel)
+        while (isActive) {
+            try {
+                val f = CompletableFuture<Void>()
+                status.portsStatus(
+                    PortsStatusRequest.newBuilder().setObserve(true).build(),
+                    object : ClientResponseObserver<PortsStatusRequest, PortsStatusResponse> {
+
+                        override fun beforeStart(requestStream: ClientCallStreamObserver<PortsStatusRequest>) {
+                            lifetime.onTerminationOrNow {
+                                requestStream.cancel(null, null)
+                            }
+                        }
+
+                        override fun onNext(ps: PortsStatusResponse) {
+                            for (port in ps.portsList) {
+                                // Avoiding undesired notifications
+                                if (ignorePorts.contains(port.localPort)) {
+                                    continue
+                                }
+
+                                val hasPreviousStatus = portsStatus.containsKey(port.localPort)
+
+                                if (!hasPreviousStatus) {
+                                    portsStatus[port.localPort] = port
+                                }
+
+                                val wasServed = portsStatus[port.localPort]?.served!!
+                                val wasExposed = portsStatus[port.localPort]?.hasExposed()!!
+                                val wasServedExposed = wasServed && wasExposed
+                                val isServedExposed = port.served && port.hasExposed()
+
+                                // If the initial update received shows that the port is served and exposed, then notify
+                                val isFirstUpdate = !hasPreviousStatus && wasServedExposed && isServedExposed
+
+                                // If the port changes its status to served and exposed, notify the user
+                                val shouldSendNotification = isFirstUpdate || !wasServedExposed && isServedExposed
+
+                                portsStatus[port.localPort] = port
+
+                                if (shouldSendNotification) {
+                                    if (port.exposed.onExposed.number == OnPortExposedAction.ignore_VALUE) {
+                                        continue
+                                    }
+
+                                    if (port.exposed.onExposed.number == OnPortExposedAction.open_browser_VALUE) {
```

Comment:
Unfortunately when the IDE is still loading, this api call to open the browser gets lost. If someone has an idea on how to delay/persist it - until the IDE is ready - let me know. Otherwise I would suggest to move forward and ask our friends @ JetBrains for suggestions.

Reply:
it sounds like we should listen for each project session (window) separately in https://github.com/gitpod-io/gitpod/blob/main/components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodClientProjectSessionTracker.kt#L20

And execute actions in client closure using:
```
ClientId.withClientId(session.clientId) {
  send notification or open browser
}
```

Reply:
it worked! thanks @akosyakov 🙌